controller/src/main/java/io/pravega/controller/task/Stream/StreamTransactionMetadataTasks.java

```diff
@@ -198,15 +207,20 @@ public Void initializeStreamWriters(final String commitStreamName, final EventSt
      * @param contextOpt optional context
      * @return true/false.
      */
-    @Task(name = "commitTransaction", version = "1.0", resource = "{scope}/{stream}/{txId}")
     public CompletableFuture<TxnStatus> commitTxn(final String scope, final String stream, final UUID txId,
                                                   final OperationContext contextOpt) {
-        final OperationContext context = getNonNullOperationContext(scope, stream, contextOpt);
+        return checkReady().thenComposeAsync(x -> {
+            final OperationContext context = getNonNullOperationContext(scope, stream, contextOpt);
+            return commitTxnBody(hostId, scope, stream, txId, context);
+        }, executor);
+    }
 
-        return execute(
-                new Resource(scope, stream, txId.toString()),
-                new Serializable[]{scope, stream, txId, null},
-                () -> commitTxnBody(scope, stream, txId, context));
+    private CompletableFuture<Void> checkReady() {
+        if (!ready) {
+            return FutureHelpers.failedFuture(new IllegalStateException(getClass().getName() + " not yet ready"));
```

Comment:
An error shall be thrown only if the object is not yet ready. We need to create this object to start all the services asynchronously from ControllerServiceStarter. However, this object shall become ready only after abort and commit streams are created by the controller event processors' bootstrap method.